#summary Using a small sample data set the main functionality of Allim is demonstrated


<wiki:toc max_depth="3" />


=Introduction=
Allim, Allelic imbalance meter, offers an integrated and user-friendly solution for measuring allele specific gene expression (ASE) within species. Allim estimates allelic imbalance in F1 hybrids. Since mapping bias is the largest problem for reliable estimates of allele specific gene expression using RNA-seq, Allim combines two different measures to account for mapping biases. First, Allim generates a polymorphism aware reference genome that accounts for the sequence variation between the alleles of both parents (or parental lines). Second, Allim includes a sequence specific simulation tool to estimate the remaining mapping bias. This estimated mapping bias is then incorporated in the statistical tests for allelic imbalance.

The pipeline requires whole transcript high throughput RNA sequencing (RNA-seq) reads of F1 hybrids. Additionally, either RNA-seq reads for both parents, genomic sequencing reads for both parents or two private genomes for both parents have to be provided in separate files. Allim was tested on Illumina paired-end RNA-seq reads but it can also handle FASTQ files from other NGS sequencing platforms. The provided parental RNA-seq libraries can be from homozygous as well as heterozygous parents (however, all heterozygous loci will be excluded during the analysis).
Allim has five modules that can be run by a single command. All input parameters can be specified in the AllimOptions file. These parameters are then used to run the complete pipeline. Allim provides two different input options:

*AllimOptions_2Pexpr:* This configuration file can be used when RNA-seq data for parent1, parent2 and F1 (hybrid) is provided. In this case Allim uses parent1 and parent2 RNA-seq reads to identify the fixed SNPs. The fixed SNPs are subsequently used to generate two “private -genomes”, one for each parent. Note, rather than RNA-seq data, also DNA reads can be provided for the two parents. 

*AllimOptions_2Pgenomes:* This configuration file can be used when the user provides RNA-seq data for the F1 hybrid and a genomic sequence for each parental line. If genomic sequences of both parents are provided, both FASTA files need to have the same size and identical FASTA IDs.



== Five Modules of Allim ==

  * Identification of fixed SNPs
  * Computer simulation of RNA-seq reads with fixed SNPs
  * Estimation of the remaining mapping bias with simulated data
  * Estimation of allele specific expression for experimental data
  * Statistical test of significant allelic imbalance

The source code and the user manual of Allim are available from http://code.google.com/p/allim/


=2. Requirements=

To use Allim pipeline, user needs to meet these requirements.
  * Linux or Macintosh OSX system or any other Unix 64 bit system with at least 4 GB of RAM and 2 CPU (processors)
  * Python 2.7.3 [http://www.python.org/download/]
  * Biopython 1.59 or higher [http://biopython.org/wiki/Download]
  * PICARD (BuildBamIndex.jar, MergeSamFiles.jar, SortSam.jar) [http://sourceforge.net/projects/picard/files/picard-tools/]
  * SAMTOOLS [http://sourceforge.net/projects/samtools/files/samtools/]
  * BedTools (bamToBed, intersectBed, sortBed) [http://code.google.com/p/bedtools/]
  * GSNAP [http://research-pub.gene.com/gmap/]
  * R version 2.15.0 [http://cran.r-project.org/]
  * NumPy [http://numpy.scipy.org/]
  * RPy 2.2.2 [http://rpy.sourceforge.net/rpy2.html]
  * R package car version 2.0-12 [http://cran.r-project.org/web/packages/car/index.html]
  * R package multcomp version 1.2-12 [http://cran.r-project.org/web/packages/multcomp/index.html]
  * Bioconductor package edgeR [http://www.bioconductor.org/packages/2.10/bioc/html/edgeR.html]
  * Bioconductor package limma [http://www.bioconductor.org/packages/release/bioc/html/limma.html]

   Note: To install R packages listed above 11-14, R >=2.15.0 should be installed.
==2.1 Operating system==
The Allim package is designed to work with a 64-bit Unix operating system with at least 4 GB of RAM and 2 CPUs (processors).

==2.2 Python installation==
Allim is developed on Python version 2.7.3. Python 2.7.3 for your operating system can be obtained from http://www.python.org/download/. After the download follow the instructions given on the web page to complete the python installation and configuration.

==2.3 Biopython installation==
Allim requires biopython for efficient fasta sequence reading, writing and other sequence manipulations. Once python is installed, biopython can be obtained and installed with the following steps:

*Step1:* Download the biopython source code from http://biopython.org/DIST/biopython-1.60.tar.gz

*Step2:* Uncompress the downloaded file with the following command:
	
{{{
tar –zxvf biopython-1.60.tar.gz
}}}

This command will return the folder/directory biopython-1.6.0

*Step3:* Enter the uncompressed folder with following command:

{{{
cd biopython-1.60
}}}

*Step4:* To install the biopython package run these two commands:

{{{
python setup.py build
python setup.py install
}}}


==2.4 NumPy installation==
NumPy is the fundamental package for scientific computing with Python. In the Allim pipeline NumPy is used to integrate sequencing errors during the simulation of RNA-seq reads. In order to install NumPy on any Unix system the following steps are required:

*Step1:* Download the NumPy package source code from http://sourceforge.net/projects/numpy/files/NumPy/1.6.0/


*Step2:* Uncompress the downloaded file with following command:
	
{{{
tar –zxvf numpy-1.6.0.tar.gz
}}}

This command will return the folder/directory numpy-1.6.0

*Step3:* Enter the uncompressed folder with following command:
	
{{{
cd numpy-1.6.0
}}}


*Step4:* To install this python package run these two commands:
	
{{{
python setup.py build
python setup.py install
}}}



==2.5 PICARD installation==
Picard comprises Java-based command-line utilities that manipulate SAM/BAM files. It is a collection of many JAVA jar files, which can be downloaded and used directly without prior installation.


*Step1:* Download the latest version of PICARD from http://sourceforge.net/projects/picard/files/picard-tools/1.75/

*Step2:* Uncompress the downloaded file with the following command:

{{{
unzip picard-tools-1.75.zip
}}}

The Allim pipeline uses three jar files: 1) _BuildBamIndex.jar_, 2) _MergeSamFiles.jar_, 3) _SortSam.jar_. Provide the full paths of these three jar files in AllimOptions run file to run the Allim pipeline. 

==2.6 SAMTOOLS installation==
SAMTOOLS provide various utilities for the manipulation of alignments in the SAM format, including sorting, merging, indexing and generating alignments in a per-position format.


*Step1:* Download latest SAMTOOLS source code from http://sourceforge.net/projects/samtools/files/samtools/


*Step2:* Uncompress the downloaded file samtools-0.1.18.tar.bz2 with following command:

{{{
tar -xvjf samtools-0.1.18.tar.bz2
}}}

This command will return the folder/directory samtools-0.1.18

*Step3:* Enter the uncompressed folder samtools-0.1.18 with following command:
	
{{{
cd samtools-0.1.18
}}}


*Step4:* To make samtools executable run the make command:
	
{{{
make
}}}

After running the make command an executable called samtools will be created. Provide the full path of the samtools executable in the AllimOptions run file to run the Allim pipeline.



==2.7 BedTools installation==
BedTools provide a set of functions to sort and intersect various genomic formats including bam files and files that contain genomic annotation such as gene position and SNP information.

*Step1:* Download BedTools from http://bedtools.googlecode.com/files/BEDTools.v2.16.2.tar.gz

*Step2:* Uncompress the downloaded file BEDTools.v2.16.2.tar.gz with the following command:

{{{
tar -xzvf BEDTools.v2.16.2.tar.gz
}}}

This command will return the folder/directory BEDTools-Version-2.16.2


*Step3:* Enter the uncompressed folder BEDTools-Version-2.16.2 with following command:
	
{{{
cd BEDTools-Version-2.16.2
}}}


*Step4:* To make BedTools executable run the make all command:
	
{{{
make all
}}}

After running the “make all” command, BEDTools-Version-2.16.2/bin sub-folder/sub-directory will be created, which contains all BedTool executables.
The Allim pipeline uses three executables: 1) _intersectBed_, 2) _sortBed_, 3) _bamToBed_. Provide the full paths of these three executables in the AllimOptions run file to run the Allim pipeline.




==2.8 GSNAP installation==
GSNAP (Genomic Short-read Nucleotide Alignment Program) is a mapper for RNA-seq data. It has the advantages that it can detect splicing events and is capable of SNP tolerant alignments (Wu & Nacu 2010).

*Step1:* Download the GSNAP source code from http://research-pub.gene.com/gmap/src/gmap-gsnap-2012-07-20.tar.gz

*Step2:* Uncompress the downloaded file “gmap-gsnap-2012-07-20.tar.gz” with following command:
{{{
tar -xzvf gmap-gsnap-2012-07-20.tar.gz
}}}


*Step3:* Enter the uncompressed folder “gmap-2012-07-20” with following command:
	
{{{
cd gmap-2012-07-20
}}}


*Step4:* To install the gmap package run these four commands:

{{{
./configure
make
make check   (optional)
make install
}}}

The above four commands will build GSNAP and other executables in “/usr/bin”

*Step5:* To check the GSNAP installations run this command:
	
{{{
gsnap –help
}}}

This command will return the detailed help manual for various GSNAP parameters if the installation was successful.

*Note:* For more detailed information how to install GSNAP please read the gmap-2012-07-20/README file.



==2.9 R installation==
The R source code specific to your operating system can be obtained from http://cran.r-project.org/. To install R, please follow the instructions provided with R.


==2.10 RPy2 installation==
RPy2 is a python class designed to do statistics programming using R in python. In order to install RPy2 on any Unix system the following steps are required:

*Step1:* Download the RPy2 package source code from http://sourceforge.net/projects/rpy/files/rpy2/2.2.x/

*Step2:* Uncompress the downloaded file with following command:
	
{{{
tar –zxvf rpy2-2.2.2.tar.gz
}}}

This command will return the folder/directory rpy2-2.2.2

*Step3:* Enter the uncompressed folder with following command:
	
{{{
cd rpy2-2.2.2
}}}


*Step4:* To install the RPy2 package run the following two commands:
	
{{{
python setup.py build
python setup.py install
}}}



==2.11 Installation of the R package _car_==

*Step1:* Download the R package car source code from http://cran.r-project.org/src/contrib/car_2.0-12.tar.gz

*Step2:* Open the terminal and run the following command:

{{{
R CMD INSTALL car_2.0-12.tar.gz
}}}



==2.12 Installation of the R package _multcomp_==

*Step1:* Download R package multcomp source code from http://cran.r-project.org/src/contrib/multcomp_1.2-12.tar.gz

*Step2:* Open the terminal and run the following command:

{{{
R CMD INSTALL multcomp_1.2-12.tar.gz
}}}



==2.13 Installation of the R package _edgeR_==

*Step1:* Download R package edgeR source code from www.bioconductor.org/packages/2.3/bioc/src/contrib/edgeR_1.0.4.tar.gz

*Step2:* Open the terminal and run following command:

{{{
R CMD INSTALL edgeR_1.0.4.tar.gz
}}}



==2.14 Installation of the R package _limma_==

*Step1:* Download R package limma source code from http://phase.hpcc.jp/mirrors/stat/R/CRAN/src/contrib/limma_2.0.2.tar.gz

*Step2:* Open the terminal and run the following command:

{{{
R CMD INSTALL limma_2.0.2.tar.gz
}}}

=Allim Installation=
The user can download the latest version of Allim from http://code.google.com/p/allim/. The file to download is called Allim_1.0.tar.gz. Move the file to an appropriate directory and run the following command to uncompress the file:
{{{
tar –zxvf Allim_1.0.tar.gz
}}}

Note that after uncompressing the tar.gz file, a new folder will be created named Allim_1.0. This directory contains the following files:


http://allim.googlecode.com/files/Allim_folder.png

=The Validation of the Installation & Sample Input Files=
To validate the installation of the Allim pipeline it can be run with a small test data set. The test data set and the corresponding Allim configuration files can be obtained from the following URLs:

*Parent1 RNA-seq:* http://allim.googlecode.com/files/parent1_RNAseq_fastq.tar.gz

*Parent2 RNA-seq:* http://allim.googlecode.com/files/parent2_RNAseq_fastq.tar.gz

*F1 RNA-seq:* http://allim.googlecode.com/files/F1_RNAseq_fastq.tar.gz

*Parent1 genome:* http://allim.googlecode.com/files/parent1_genome.fa

*Parent2 genome:* http://allim.googlecode.com/files/parent2_genome.fa

*Reference fasta file:* http://allim.googlecode.com/files/reference.fa

*Gene annotation file:* http://allim.googlecode.com/files/reference.gtf

*Allim configuration file:* http://allim.googlecode.com/files/AllimOptions_2Pexpr

*Allim configuration file:* http://allim.googlecode.com/files/AllimOptions_2Pgenomes



*Note:* You either need to use the parental RNA-seq or the parental genomic data. The option you choose depends on the type of input data you have for your own analysis.

Create a folder named *“test_data”* in the Allim_1.0 directory and download all the above test data files into this folder, unzip and extract the compressed files. Move the *AllimOptions_2Pexpr*  and *AllimOptions_2Pgenomes* files into the directory *Allim_1.0* directly. The Allim pipeline can then be run on the test data set with the following two steps:


*1.* Open the terminal and enter the Allim_1.0 directory.
cd Allim_1.0

*2.* Run the Allim pipeline with the test data set via the following command (the type of the AllimOptions file that is used depends on the type of your input data):
{{{
python 01-allim.py --option-file AllimOptions_2Pexpr
OR
python 01-allim.py --option-file AllimOptions_2Pgenomes
}}}

To get help on how to run Allim and required parameters enter: 
{{{
	python 01-allim.py –help
}}}



=Benchmark for Allim pipeline=
We have run Allim pipeline with following data sets.

*RNA-Seq reads:*

Parent1 reads: 2 million 100 bp paired-end reads with 68 bp average insert size
Parent2 reads: 2 million 100 bp paired-end reads with 78 bp average insert size
Hybrid reads: 2 million 100 bp paired-end reads with 98 bp average insert size

*Computer simulated reads:*

Parent1 reads: 10.26 million 100 bp paired-end reads with 78 bp insert size
Parent2 reads: 10.26 million 100 bp paired-end reads with 78 bp insert size

||*S.No.*||*Module*||*Time (min:sec)*||
||1.||Finding fixed SNP||95:00||
||2.||Computer simulation of RNA-Seq reads||20:00||
||3.||Estimating Allele Specific Expression with experimental data||24:00||
||4.||Estimating Allele Specific Expression with simulated data||34:00||
||5||Statistical test of significant allelic imbalance||3:00||

*Table1: Benchmarks for processing time for each module of Allim*

*Note:* All benchmarks have been done on a Mac OS X 10.6.8, 2x2.8 GHz Quad-Core Intel Xeon, with 4 GB of RAM using 4 processors (CPU)

=Validation=
We used RNA-Seq data from females and males of two different Drosophila pseudoobscura isofemale lines (*_Table2_*) and ran our Allim pipleline after pooling the reads from both libraries. In contrast to experiments measuring allelic imbalance, the origin of each read is known.Our results show that 99% (*_Table3_*) of the informative reads are assigned to correct parental allele.

*Table2: RNA-Seq data used for validation*

||Samples||Reads (in million)||Insert Size (bp)||
||ps94 males||79.21||78||
||ps94 females||80.00||78||
||ps88 males||79.21||128||
||ps88 females||80.00||68||

Further more validated Allim via simulations of RNA-Seq of both parental alleles (i.e.: 2x 100bp paired ends with 78bp insert size). For both genomes the same number of reads covering a polymorphic site were generated. Thus, in absence of a mapping bias, all genes should have an expression ratio of one for every gene. This result also shows about 99.99% (*_Table3_*) informative reads are assigned to correct parental allele. 

*Table3 Accuracy of Allim pipeline*
||Dataset||strainA (%)||strainB (%)||
||Females||99.97||98.96||
||Males||99.96||98.60||
||Simulation||99.99||99.99||



==Female data==
Expression of each gene in strainA and strainB plotted as given below. Correlation coefficient (R^2)=0.948; P-value < 2.2e-16.

http://allim.googlecode.com/files/female-expr.jpg

==Male data==
Expression of each gene in strainA and strainB plotted as given below. Correlation coefficient (R^2)=0.959; P-value < 2.2e-16.

http://allim.googlecode.com/files/male-expr.jpg

==Simulation data==
Expression of each gene in strainA and strainB plotted as given below. Correlation coefficient (R^2)=0.992; P-value < 2.2e-16.

http://allim.googlecode.com/files/simulated-expr.jpg

== Gene expression ratio==
The expression ratio for each expressed gene is calculated as strainA/strainB for both data (Females and Males). In given boxplot the log10 ratio is shown on y-axis.

 
http://allim.googlecode.com/files/expr-ratio-boxplot.jpg